packages/vite/src/node/optimizer/optimizer.ts

```diff
@@ -664,30 +676,65 @@ async function createDepsOptimizer(
     }
   }
 
-  const runOptimizerIfIdleAfterMs = 50
+  let crawlEndFinder = setupOnCrawlEnd(onCrawlEnd)
+
+  // Called during buildStart at build time, when build --watch is used.
+  // On dev, onCrawlEnd is called once. On build, we call it when all the
+  // user modules are processed for every rebuild.
+  function resetRegisteredIds() {
+    crawlEndCalled = false
+    crawlEndFinder.cancel()
+    crawlEndFinder = setupOnCrawlEnd(onCrawlEnd)
+  }
+
+  function registerWorkersSource(id: string) {
+    crawlEndFinder.registerWorkersSource(id)
+  }
+  function delayDepsOptimizerUntil(id: string, done: () => Promise<any>) {
+    if (!crawlEndCalled && !depsOptimizer.isOptimizedDepFile(id)) {
+      crawlEndFinder.delayDepsOptimizerUntil(id, done)
+    }
+  }
+  function ensureFirstRun() {
+    if (!firstRunCalled) crawlEndFinder.ensureFirstRun()
+  }
+}
+
+const runOptimizerIfIdleAfterMs = 50
 
+function setupOnCrawlEnd(onCrawlEnd: () => void): {
+  ensureFirstRun: () => void
+  registerWorkersSource: (id: string) => void
+  delayDepsOptimizerUntil: (id: string, done: () => Promise<any>) => void
+  cancel: () => void
+} {
   let registeredIds: { id: string; done: () => Promise<any> }[] = []
-  let seenIds = new Set<string>()
-  let workersSources = new Set<string>()
-  const waitingOn = new Set<string>()
+  const seenIds = new Set<string>()
+  const workersSources = new Set<string>()
+  const waitingOn = new Map<string, () => void>()
   let firstRunEnsured = false
+  let crawlEndCalled = false
 
-  function resetRegisteredIds() {
-    registeredIds = []
-    seenIds = new Set<string>()
-    workersSources = new Set<string>()
-    waitingOn.clear()
-    firstRunEnsured = false
+  let cancelled = false
+  function cancel() {
+    cancelled = true
+  }
+
+  function callOnCrawlEnd() {
+    if (!cancelled && !crawlEndCalled) {
+      crawlEndCalled = true
+      onCrawlEnd()
+    }
   }
 
   // If all the inputs are dependencies, we aren't going to get any
   // delayDepsOptimizerUntil(id) calls. We need to guard against this
   // by forcing a rerun if no deps have been registered
   function ensureFirstRun() {
-    if (!firstRunEnsured && !firstRunCalled && registeredIds.length === 0) {
+    if (!firstRunEnsured && seenIds.size === 0) {
       setTimeout(() => {
-        if (!closed && registeredIds.length === 0) {
-          onCrawlEnd()
+        if (seenIds.size === 0) {
+          callOnCrawlEnd()
```

Comment:
`registeredIds.length === 0` to `seenIds.size === 0` is also needed after #12609